File: js/primes.js
const { performance, PerformanceObserver } = require('perf_hooks')

const perfObserver = new PerformanceObserver(() => {})

perfObserver.observe({ entryTypes: ["measure"], buffer: true })
performance.mark('start')

let max = 1000;
let primes = [];

if (process.argv.length > 2) {
    const maybeInt = parseInt(process.argv[2])
    if (!isNaN(maybeInt)) {
        max = maybeInt
    }
}


for (let i = 1; i <= max; i++) {
    let out = `Is ${i} prime? `
    let isPrime = true;
    for (p of primes) {
        if (p == 1 || p == i) {
            continue
        }
        if (i % p == 0) {
            isPrime = false;
            break
        }
    }
    if (isPrime) {
        primes.push(i)
        out += 'yes\r'
    }
    else {
        out += 'no \r'
    }
    process.stdout.write(out)
}

console.log(`\nCount: ${primes.length}`)

performance.mark('end')
const perfResult = performance.measure('runtime', 'start', 'end')


console.log(`Time (ns): ${(perfResult.duration * 1_000_000.0).toFixed(0)}`)
console.log(`Time (µs): ${(perfResult.duration * 1_000.0).toFixed(3)}`)
console.log(`Time (ms): ${perfResult.duration.toFixed(3)}`)
console.log(`Time (s):  ${(perfResult.duration / 1_000.0).toFixed(4)}`)

File: python/primes.py
import time
import sys

def main():
    primes = []
    largest_number = 1000
    if len(sys.argv) > 1 and len(sys.argv[1]) != 0:
        largest_number = int(sys.argv[1])
    i = 1
    while i <= largest_number:
        print(f"Is {i} prime?\t", end='')
        is_prime = True
        for p in primes:
            if p == 1 or p == i:
                continue
            if i % p == 0:
                is_prime = False    
        if is_prime:
            print("yes", end='\r', flush=True)
            primes.append(i)
        else:
            print("no ", end='\r', flush=True)
        i += 1
    print(f"\nCount: {len(primes)}")

if __name__ == "__main__":
    start = time.time_ns()
    main()
    end = time.time_ns()
    td = end - start
    print(f"Time (ns): {td}")
    print("Time (µs): {0:.3f}".format(td / 1_000))
    print("Time (ms): {0:.3f}".format(td / 1_000_000))
    print("Time (s):  {0:.4f}".format(td / 1_000_000_000))

